src/nvim/decoration_provider.c

```diff
@@ -47,11 +48,32 @@ static bool decor_provider_invoke(NS ns_id, const char *name, LuaRef ref, Array
   return false;
 }
 
+bool decor_provider_invoke_spell(win_T *wp, linenr_T lnum, colnr_T col, char **err)
+{
+  bool ret = false;
+  for (size_t i = 0; i < kv_size(decor_providers); i++) {
+    DecorProvider *p = &kv_A(decor_providers, i);
+    if (!p->active) {
+      continue;
+    }
+
+    if (p->is_spell != LUA_NOREF) {
+      MAXSIZE_TEMP_ARRAY(args, 4);
+      ADD_C(args, INTEGER_OBJ(wp->handle));
+      ADD_C(args, INTEGER_OBJ(wp->w_buffer->handle));
+      ADD_C(args, INTEGER_OBJ(lnum - 1));
+      ADD_C(args, INTEGER_OBJ(col - 1));
+      ret |= decor_provider_invoke(p->ns_id, "spell", p->is_spell, args, true, err);
+    }
+  }
+  return ret;
+}
```

Comment:
This isn't what we talked about on matrix. This completely ignores anything about extmarks.

Reply:
I think this is also very expensive since you need to call the callback for every column. With the way we had it before, even though the PR didn't, you only had to call `on_line` per line.

Reply:
Maybe we can do the following then: `on_spell` will be called once per line considered, and is expected to set ephemeral spellchecking extmarks for this line ?

Reply:
Given the output of the test though, this is not _that_ expensive and this is only called when a _badly spelled_ word is found.

Reply:
I think we can be a little more general. Instead of once per line, we can provide a start position and an end line. If `end < start` then it means go backwards. When called it basically means, "I need to find the closest spelling error in this direction, populate me some extmarks so I can find them".

However, in practice, we will only call it per line until we can refactor `spell_move_to` more. This will slightly couple the ts highlighter and `spell_move_to` but that's fine since they are in the same codebase.

> Given the output of the test though, this is not that expensive and this is only called when a badly spelled word is found.

This assumes the implementation stays as it is. What if we wanted to find the marks first and then spell check?

Reply:
Also to note, the end position is only necessary to provide flexibility in the implementation. An ideal implementation will just provide the start or end of the buffer.

Reply:
we could merge this more or less as is, but rename the callback `_is_spell` to indicate that the interface is not finalized and will likely be changed in 0.9.

Reply:
I've implemented what I've described above and renamed it as `_on_spell_nav`.